src/classes/VOL_CTRL_VolunteersCampaignWizard.cls

```diff
@@ -117,267 +119,263 @@ public with sharing class VOL_CTRL_VolunteersCampaignWizard {
+        listStrFieldsShift.addAll(mapS.keySet());
+
+        // Ensure the user has access to the object and fields before querying
+        UTIL_Describe.checkReadAccess(UTIL_Describe.StrTokenNSPrefix('Volunteer_Shift__c'),
+            new Set<String>(listStrFieldsShift));
+
+        // find the existing shifts for those jobs
+        // we need to use dynamic soql to pick up all custom fields we don't know about
+        strSoql = 'select ';
+        strComma = '';
+        for (string strF : listStrFieldsShift) {
+            strSoql += strComma + strF;
+            strComma = ', ';
+        }
+        strSoql += ' from Volunteer_Shift__c where Volunteer_Job__c in :listJobsExisting order by Start_Date_Time__c';
+        listShiftsExisting = Database.Query(strSoql);
+
+        // clone the jobs
+        for (Volunteer_Job__c job : listJobsExisting) {
+            Volunteer_Job__c jobNew = new Volunteer_Job__c();
+
+            //  now copy over all the non-null fields
             for (string strF : listStrFieldsJob) {
-                strSoql += strComma + strF;
-                strComma = ', ';
+                if (strF != 'Id' && job.get(strF) != null) {
+                    try {
+                        jobNew.put(strF, job.get(strF));
+                    } catch (exception ex) {
+                        // catch errors from readonly fields, and continue.
+                    }
+                }
             }
-            strSoql += ' from Volunteer_Job__c where Campaign__c = :campaignIdClone order by CreatedDate';    
-            listJobsExisting = Database.Query(strSoql); 
+            jobNew.Campaign__c = cmpVols.Id;
+            listJobs.add(jobNew);
+        }
+
+        // Ensure the user has access to the object and fields before querying
+        UTIL_Describe.checkObjectCreateAccess(UTIL_Describe.StrTokenNSPrefix('Volunteer_Job__c'));
+        insert listJobs;
+
+        // create a map of existing JobId to new JobId
+        map<ID, ID> mapJobId = new map<ID, ID>();
+        for (Integer i = 0; i < listJobsExisting.size(); i++) {
+            mapJobId.put(listJobsExisting[i].Id, listJobs[i].Id);
+        }
+
+        // figure out our base date to use.
+        Date dt = cmpShiftDates.EndDate;
+        if (dt == null) dt = cmpVols.StartDate;
+        if (dt == null) dt = System.Now().date();
+
+        // figure out date offset
+        integer cDays;
+        if (cmpShiftDates.StartDate == null) {
+            cDays = -1; // don't offset, use fixed dt.
+        } else {
+            cDays = cmpShiftDates.StartDate.daysBetween(dt);
+        }
+
+        // clone the shifts
+        for (Volunteer_Shift__c shift : listShiftsExisting) {
+            Volunteer_Shift__c shiftNew = new Volunteer_Shift__c();
+
+            //  now copy over all the non-null fields
+            for (string strF : listStrFieldsShift) {
+                if (strF != 'Id' && shift.get(strF) != null) {
+                    try {
```

Comment:
Same questions as above regarding the try/catch block and exception handling.

Reply:
as the comments says, to deal with readonly fields (formulas) that exist on the object.